simulation/model.py

```diff
@@ -0,0 +1,102 @@
+import datetime
+from collections import Counter
+from typing import Dict, Iterator
+
+import pandas as pd
+
+from abm import AgentBasedModel, UnorderedSpace
+from simulation.agents import Household
+from simulation.collectors import get_agent_collectors, model_collectors
+from simulation.constants import (
+    BuiltForm,
+    Epc,
+    HeatingSystem,
+    OccupantType,
+    PropertyType,
+)
+
+
+class CnzAgentBasedModel(AgentBasedModel):
+    def __init__(
+        self,
+        start_datetime,
+        step_interval,
+    ):
+        self.start_datetime = start_datetime
+        self.step_interval = step_interval
+        self.current_datetime = start_datetime
+
+        super().__init__(UnorderedSpace())
+
+    @property
+    def current_agent_count(self) -> Dict:
+        return Counter([agent.__class__ for agent in self.space.agents])
+
+    @property
+    def current_heating_system_count(self) -> Dict[HeatingSystem, int]:
+        return Counter(
+            [
+                agent.heating_system
+                for agent in self.space.agents
+                if isinstance(agent, Household)
+            ]
+        )
+
+    def step(self):
+        if self.current_datetime == self.start_datetime:
+            self.initial_heating_system_count = self.current_heating_system_count
+        self.current_datetime += self.step_interval
+
+
+def parse_string_interval(string_range: str) -> pd.Interval:
+    min_value, max_value = [int(i) for i in string_range.split("-")]
+    return pd.Interval(min_value, max_value, closed="both")
+
+
+def create_households(
+    num_households: int,
+    household_distribution: pd.DataFrame,
+) -> Iterator[Household]:
+    household_distribution["construction_year_band"] = household_distribution[
+        "construction_year_band"
+    ].apply(parse_string_interval)
+
+    households = household_distribution.sample(num_households, replace=True)
+    for household in households.itertuples():
+        yield Household(
+            location=household.location,
+            property_value=household.property_value,
+            floor_area_sqm=household.floor_area_sqm,
+            off_gas_grid=household.off_gas_grid,
+            construction_year_band=household.construction_year_band,
```

Comment:
I think we should change the construction year band to either:
1. A single int value `construction_year: int`, of the year it was constructed, and do the preprocessing from the string to that value at this stage
2. A `enum.Enum` for `ConstructionYearBand` which has all the property year archetypes we are using in there. We should map the string onto that enum at this stage, too. Then `construction_year_band: ConstructionYearBand` and we can work with it a bit easier.

I would be in favour of 2.

Reply:
Note: if we did 2 we can remove all the `pd.Interval` stuff we have, and preprocessing functions etc.

Reply:
I don't mind - I feel that sampling uniformly from the EPC construction year bands results is unlikely to be the best approach anyway (there were probably some property booms within these sub-periods, not uniform in count). It would remove any ability to cap heating system age by construction year though (which I removed for this first PR, and I think is fine to continue to omit).